### PR TITLE
CORS-3073: azure: Allow for installs in existing vnets

### PR DESCRIPTION
**Pull request by rna-afk**
Adding the option to allow for existing vnets installation in CAPZ.

### CODE REVIEW

File: pkg/asset/manifests/azure/cluster.go [hunk 1]
```diff
@@ -86,7 +86,8 @@ func GenerateClusterAssets(installConfig *installconfig.InstallConfig, clusterID
 					PrivateDNSZoneName: installConfig.Config.ClusterDomain(),
 				},
 				Vnet: capz.VnetSpec{
-					ID: installConfig.Config.Azure.VirtualNetwork,
+					ResourceGroup: installConfig.Config.Azure.NetworkResourceGroupName,
+					Name:          installConfig.Config.Azure.VirtualNetwork,
```
Comment (sadasu) [inline pkg/asset/manifests/azure/cluster.go:90]: Looking at https://github.com/openshift/installer/blob/master/data/data/install.openshift.io_installconfigs.yaml#L2701, it appears that VirtualNetwork is an optional config. So, is it possible that we are creating capz.VnetSpec{} with Name as " " ?
Similar comment about https://github.com/openshift/installer/blob/master/data/data/install.openshift.io_installconfigs.yaml#L2663?

Comment (rna-afk) [reply pkg/asset/manifests/azure/cluster.go:90]: CAPZ doesn't consider it if it's left blank. Tested it and it defaults to the `-vnet` default and creates it.

**Comment by jhixson74** [comment]
In testing this, I am seeing: 

```
jhixson@fedora:~/test/azure.pre-existing-vnet$ oc get clusteroperator ingress
NAME      VERSION   AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE
ingress             False       True          True       32m     The "default" ingress controller reports Available=False: IngressControllerUnavailable: One or more status conditions indicate unavailable: LoadBalancerReady=False (SyncLoadBalancerFailed: The service-controller component is reporting SyncLoadBalancerFailed events like: Error syncing load balancer: failed to ensure load balancer: nsg "jhfoo-879p4-nsg" not found...

```
Looking at the cluster-api code, it looks like you can specify the name of an existing network security group: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/122ef5cb227e4941899d2946e446aa4a91391d8c/api/v1beta1/azurecluster_default.go#L178

File: pkg/infrastructure/azure/azure.go [hunk 2]
```diff
@@ -108,6 +109,51 @@ func (p *Provider) PreProvision(ctx context.Context, in clusterapi.PreProvisionI
...
+			armnetwork.SecurityGroup{
+				Location: to.Ptr(platform.Region),
+				Tags:     tags,
+				Properties: &armnetwork.SecurityGroupPropertiesFormat{
+					SecurityRules: []*armnetwork.SecurityRule{
+						{
+							Name: to.Ptr("apiserver_in"),
+							Properties: &armnetwork.SecurityRulePropertiesFormat{
+								Protocol:                 to.Ptr(armnetwork.SecurityRuleProtocolTCP),
+								SourceAddressPrefix:      to.Ptr("*"),
```
Comment (patrickdillon) [inline pkg/infrastructure/azure/azure.go:133]: For private clusters, we should restrict this to the machine network cidr

File: pkg/infrastructure/azure/azure.go [hunk 3]
```diff
@@ -11,7 +11,11 @@ import (
 
 	"github.com/Azure/azure-sdk-for-go/sdk/azcore/arm"
 	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+<<<<<<< HEAD
 	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/authorization/armauthorization/v3"
+=======
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
+>>>>>>> 33b11b4f0c (azure: Allow for installs in existing vnets)
```
Comment (jhixson74) [inline pkg/infrastructure/azure/azure.go:18]: Looks like your rebase didn't fully complete.

**Review by patrickdillon — COMMENTED** [review]
So we discovered that by setting the vNet ID capz considers the vnet "not managed" and will not create security groups or route tables, as we wanted. But I'm a bit confused as to why we're not hitting the -nsg not found error that john described: https://github.com/openshift/installer/pull/8662#issuecomment-2204476577 

To get around this in TF we create a dummy NSG.

**Comment by patrickdillon** [comment]
I created https://github.com/openshift/release/pull/54269 which would allow us to test this PR in ci

**Comment by patrickdillon** [comment]
The [CI test is failing](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_installer/8662/pull-ci-openshift-installer-master-altinfra-e2e-azure-ovn-shared-vpc/1811544120796123136) as expected with not being able to find the nsg:  

```
level=error msg=Cluster operator ingress Available is False with IngressUnavailable: The "default" ingress controller reports Available=False: IngressControllerUnavailable: One or more status conditions indicate unavailable: LoadBalancerReady=False (SyncLoadBalancerFailed: The service-controller component is reporting SyncLoadBalancerFailed events like: Error syncing load balancer: failed to ensure load balancer: nsg "ci-op-k9x3yf60-51127-bzngw-nsg" not found
```

In TF we get around this by creating a dummy NSG (the NSG does nothing) for the ingress controller:

https://github.com/openshift/installer/blob/36bbe11d2c1b1aa0891627265e69c59d82f4ba78/data/data/azure/vnet/nsg.tf#L1-L6

[The docs say](https://docs.openshift.com/container-platform/4.15/installing/installing_azure/installing-azure-vnet.html):

> Because cluster components do not modify the user-provided network security groups, which the Kubernetes controllers update, a pseudo-network security group is created for the Kubernetes controller to modify without impacting the rest of the environment.

File: pkg/asset/manifests/azure/cluster.go [hunk 4]
```diff
@@ -86,6 +91,13 @@ func GenerateClusterAssets(installConfig *installconfig.InstallConfig, clusterID
 					PrivateDNSZoneName: installConfig.Config.ClusterDomain(),
 				},
 				Vnet: capz.VnetSpec{
+					ResourceGroup: installConfig.Config.Azure.NetworkResourceGroupName,
+					Name:          installConfig.Config.Azure.VirtualNetwork,
+					// The ID is set to virtual network here for existing vnets here. This is to force CAPZ to consider this resource as
+					// "not managed" which would prevent the creation of an additional nsg and route table in the network resource group.
+					// The ID field is not used for any other purpose in CAPZ except to set the "managed" status.
+					// See https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/scope/cluster.go#L585
+					// https://github.com/kubernetes-sigs/cluster-api-provider-azure/commit/0f321e4089a3f4dc37f8420bf2ef6762c398c400
```
Comment (jhixson74) [inline pkg/asset/manifests/azure/cluster.go:100]: Shouldn't this only be set when installing into existing vnets? If this is always set, capz will always mark it as managed, right?

Comment (rna-afk) [reply pkg/asset/manifests/azure/cluster.go:100]: If VirtualNetwork isn't set in the install config, the ID is set to "" which is the same as not setting it.

Comment (jhixson74) [reply pkg/asset/manifests/azure/cluster.go:100]: Makes sense. I was under the impression it was not empty.

**Pull request merged**